### PR TITLE
Add new data source turbot_control

### LINKED ISSUE #32: new data source: turbot_control
This would allow you to query the status of a turbot control as part of a stack, to make decisions for resource creation

### PR DESCRIPTION
**Pull request by RupeshPatil20**
Closes #32

### CODE REVIEW

File: turbot/data_source_turbot_control.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	if controlUri != "" && resourceId != "" {
+		args = fmt.Sprintf(`uri: %s, resourceId: %s`, controlUri, resourceId)
+	} else {
+		args = fmt.Sprintf(`id: %s`, controlId)
+	}
+	control, err := client.ReadControl(args)
+	if err != nil {
+		if apiClient.NotFoundError(err) {
+			// setting was not found - clear id
+			d.SetId("")
```
Comment (kaidaguerre) [inline turbot/data_source_turbot_control.go:64]: if the error is not found, we should not return an error right, we should just set the id to empty?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]
looks good (small changes aside) - before releasing let's verify the required fields.

File: turbot/data_source_turbot_control.go (hunk 1 above)
Comment (RupeshPatil20) [reply turbot/data_source_turbot_control.go:64]: As it is a data source, we need to return error.

File: website/docs/d/control.html.markdown [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Argument Reference
+
+* `id` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `uri` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `resource` - (Optional) The unique ID of the resource at the level of which the information needs to be fetched.
+
+## Attributes Reference
+
+* `state` - The final state of the set control.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: The state of the control.

File: website/docs/d/control.html.markdown [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+* `reason` - Message explaining the state of the set control.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: remove word 'set'

File: website/docs/d/control.html.markdown [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+* `details` - Additional information regarding the set control.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: Additional information regarding the control state.

File: website/docs/d/control.html.markdown [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+* `tags` - User defined way of logically grouping resources.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: Tags set on the control.

File: website/docs/d/control.html.markdown [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+* `turbot` - JSON representation of turbot data of the resource.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: (I wonder is we should remove turbot data)
```
`turbot` - Turbot metadata for the control

File: website/docs/d/control.html.markdown [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  value = "${data.turbot_control_value.example}".status
+}
+```
+
+## Argument Reference
+
+* `id` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `uri` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `resource` - (Optional) The unique ID of the resource at the level of which the information needs to be fetched.
+
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown:45]: add a note to say either id or uri and resource must be specified

File: turbot/data_source_turbot_control.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+				},
+			},
+		},
+	}
+}
+
+func dataSourceTurbotControlRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	controlId := d.Get("id").(string)
+	controlUri, controlOk := d.GetOk("uri")
```
Comment (kaidaguerre) [inline turbot/data_source_turbot_control.go]:
```
        controlId, controlIdSet := d.GetOk("id")
	controlUri, controlUriSet := d.GetOk("uri")
	resourceId, resourceIdSet := d.GetOk("resource")
		
	var args string
       // validate parameters and build mutation arguments 
	if controlIdSet {
		if  controlUriSet || resourceIdSet{
			return fmt.Errorf("if controlId is set, controlUri and resourceId must not be set")
		}
		args = fmt.Sprintf(`id: "%s"`, controlId)
	} else {
		if !controlUriSet || !resourceIdSet {
			return fmt.Errorf("either controlId or controlUri AND resourceId must not be set")
		}
		args = fmt.Sprintf(`uri: %s, resourceId: %s`, controlUri, resourceId)
	}
```

File: website/docs/d/control.html.markdown [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+```hcl
+data "turbot_control" "test" {
+  id      = "tmod:@turbot/aws-s3#/control/types/bucketVersioning"
+}
+
+output "json" {
+  value = "${data.turbot_control.test}".status
+}
+```
+Here is another example wherein, we can fetch control data using uri and resource.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: using the control type uri and resource id

File: website/docs/d/control.html.markdown [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+output "json" {
+  value = "${data.turbot_control_value.example}".status
+}
+```
+
+## Argument Reference
+
+* `id` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `uri` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `resource` - (Optional) The unique ID of the resource at the level of which the information needs to be fetched.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: the unique identifier of the resource which the control is targetting

File: website/docs/d/control.html.markdown [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+output "json" {
+  value = "${data.turbot_control_value.example}".status
+}
+```
+
+## Argument Reference
+
+* `id` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `uri` - (Optional) The unique identifier of the control for which the value needs to be extracted.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: the type of the control

File: website/docs/d/control.html.markdown [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+output "json" {
+  value = "${data.turbot_control_value.example}".status
+}
+```
+
+## Argument Reference
+
+* `id` - (Optional) The unique identifier of the control for which the value needs to be extracted.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: The id of the control.

File: website/docs/d/control.html.markdown [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+* `uri` - (Optional) The unique identifier of the control for which the value needs to be extracted.
+* `resource` - (Optional) The unique ID of the resource at the level of which the information needs to be fetched.
+
+-> **Note:** It's recommended to either specify id of the control or uri and resource both.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: **Note:** You must specify either the control `id` or the control type `uri` AND the `resource`.

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: turbot/data_source_turbot_control.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+		}
+		args = fmt.Sprintf(`id: "%s"`, controlId)
+	} else {
+		if !controlTypeSet || !resourceIdSet {
+			return fmt.Errorf("either controlId or controlType AND resourceId must not be set")
+		}
+		args = fmt.Sprintf(`uri: %s, resourceId: %s`, controlType, resourceId)
+	}
+
+	if args == "" {
```
Comment (kaidaguerre) [inline turbot/data_source_turbot_control.go]: why is this needed?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: website/docs/d/control.html.markdown [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+output "json" {
+  value = "${data.turbot_control.test}".status
+}
+```
+Here is another example wherein, we can fetch control data using the control type uri and resource id.
+
+```hcl
+data "turbot_control_value" "example" {
+  uri      = "tmod:@turbot/aws-ec2#/control/types/instanceDiscovery"
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: type

File: website/docs/d/control.html.markdown [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+```hcl
+data "turbot_control" "test" {
+  id      = "tmod:@turbot/aws-s3#/control/types/bucketVersioning"
+}
+
+output "json" {
+  value = "${data.turbot_control.test}".status
+}
+```
+Here is another example wherein, we can fetch control data using the control type uri and resource id.
```
Comment (kaidaguerre) [inline website/docs/d/control.html.markdown]: using the control type and target resource.

File: turbot/data_source_turbot_control.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+func dataSourceTurbotControlRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	controlId, controlIdSet := d.GetOk("id")
+	controlType, controlTypeSet := d.GetOk("type")
+	resourceId, resourceIdSet := d.GetOk("resource")
+	var args string
+
+	if controlIdSet {
+		if controlTypeSet || resourceIdSet {
+			return fmt.Errorf("if controlId is set, controlType and resourceId must not be set")
```
Comment (kaidaguerre) [inline turbot/data_source_turbot_control.go]: `if id is set, type and resource must not be set`

File: turbot/data_source_turbot_control.go [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+		}
+		args = fmt.Sprintf(`id: "%s"`, controlId)
+	} else {
+		if !controlTypeSet || !resourceIdSet {
+			return fmt.Errorf("either controlId or controlType AND resourceId must not be set")
```
Comment (kaidaguerre) [inline turbot/data_source_turbot_control.go]: `"either id or type AND resource must not be set"`

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: turbot/data_source_turbot_control.go [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+	var args string
+
+	if controlIdSet {
+		if controlTypeSet || resourceIdSet {
+			return fmt.Errorf("if 'id' is set, 'type' and 'resource' must not be set")
+		}
+		args = fmt.Sprintf(`id: "%s"`, controlId)
+	} else {
+		if !controlTypeSet || !resourceIdSet {
+			return fmt.Errorf("either 'id' or 'type' AND 'resource' must not be set")
```
Comment (kaidaguerre) [inline turbot/data_source_turbot_control.go]: `"either 'id' or 'type' AND 'resource' must be set")
